### PR TITLE
refactor: Enhance ConvertTool with flexible int conversion

### PR DESCRIPTION
**Pull request by leozeli**
### Summary
Enhanced the [`ConvertTool`](tools.go:159) function to automatically handle string-to-int conversions for integer parameters, improving compatibility with MCP clients that may send numeric values as strings.

#### Core Implementation
- **New function [`unmarshalWithIntConversion`](tools.go:77)**: Intelligently unmarshals JSON data while converting string values to int for integer-typed fields
  - Handles both direct `int` fields and `*int` pointer fields
  - Supports positive, negative, and zero values
  - Provides clear error messages for invalid conversions
  
- **Updated [`ConvertTool`](tools.go:220)**: Modified to use [`unmarshalWithIntConversion`](tools.go:77) instead of standard `json.Unmarshal`, enabling flexible parameter handling

#### Testing
- **New test file [`tools_string_to_int_test.go`](tools_string_to_int_test.go:1)**: Comprehensive test coverage including:
  - String-to-int conversion for regular int fields
  - Native int value handling (backward compatibility)
  - Mixed string and int values
  - Optional int pointer fields
  - Zero and negative number handling
  - Error cases for invalid conversions
  - End-to-end tool handler integration tests

### Benefits
- **Improved Client Compatibility**: MCP clients can send int parameters as either strings or native numbers
- **Backward Compatible**: Existing tools continue to work with native int values
- **Type Safety**: Validation ensures only valid numeric strings are converted, with clear error messages
- **Zero Runtime Overhead**: Conversion happens during unmarshaling with no additional performance impact

### Testing
All tests pass with comprehensive coverage of edge cases:
```bash
✅ String to int conversion
✅ Native int values (backward compatibility)  
✅ Mixed string/int values
✅ Optional pointer fields
✅ Zero and negative numbers
✅ Invalid conversion error handling
```

### Commit
- `42f24d4` - refactor: Enhance ConvertTool with flexible int conversion
#69 

---

> [!NOTE]
> **Medium Risk**
> Changes request argument unmarshaling for all tools created via `ConvertTool`, which could subtly alter validation/error behavior for integer fields and edge-case inputs. Covered by extensive unit tests, but still impacts a central request path.
> 
> **Overview**
> `ConvertTool` now unmarshals tool `arguments` via `unmarshalWithIntConversion`, allowing integer-typed params (including embedded and pointer integer fields) to be provided as either JSON numbers or quoted strings.
> 
> Adds reflection-based detection of integer JSON field names and a retry path that strips quotes for those fields only, plus a comprehensive new test suite covering mixed types, embedded structs, overflow/invalid values, and an end-to-end tool call. Updates `go.sum` to drop `github.com/prometheus/alertmanager v0.30.0` in favor of `v0.30.1`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 882715683718d0b36dc0082e895ff0ca710da675. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by sd2k** [comment]
Hi, can you please describe the problem you're encountering which requires this fix?

**Comment by leozeli** [comment]
> Hi, can you please describe the problem you're encountering which requires this fix?

Issue Draft
Title: Model passes strings for integer parameters in tool calls, causing JSON unmarshal errors (Go)

Description:
When using the claude-sonnet-4-5-20250929-v1 model to call MCP tools, it consistently formats numerical arguments as strings in the JSON payload. This causes strict-typed backends (like Go) to throw errors during JSON unmarshaling.

Even when explicitly emphasizing the expected integer/number types in the system prompt and tool schema descriptions, the model continues to wrap numbers in quotes.

Error Example:

Plaintext
Error executing MCP tool:
MCP error -32603: unmarshal args: json: cannot unmarshal string into Go struct field ListDatasourcesParams.limit of type int
Payloads Generated by the Model:

Example 1: list_datasources

JSON
{
  "limit": "100",
  "offset": "0",
  "type": "prometheus"
}
(Expected limit and offset to be raw numbers: 100, 0)

Example 2: query_prometheus

JSON
{
  "datasourceUid": "u1wflDBO",
  "endTime": "now",
  "expr": "rate(container_network_receive_bytes_total{namespace=\"odp-platform\",pod=~\"notice-center-dev-v1.*\"}[5m]) / 1024 / 1024",
  "queryType": "range",
  "startTime": "now-1h",
  "stepSeconds": "60"
}
(Expected stepSeconds to be a raw number: 60)

Current Workaround:
Because prompting does not resolve the issue, I had to implement type compatibility on the tool execution side. I modified my Go codebase to handle these fields flexibly (e.g., accepting them as strings and parsing them to ints manually, or using custom UnmarshalJSON logic) to prevent the tool calls from failing.

Expected Behavior:
The model should strictly respect the type definitions provided in the MCP tool schema and output native JSON numbers for integer/float fields rather than casting them to strings.

**Comment by sd2k** [comment]
Thanks, that does seem like a legitimate problem, I agree. Rather than implementing this custom unmarshaling though I think we could make use of `mapstructure` and `WeaklyTypedInput` (https://pkg.go.dev/github.com/go-viper/mapstructure/v2#example-Decode-WeaklyTypedInput) which should massively reduce the amount of custom code we need to maintain here.

**Comment by leozeli** [comment]
> Thanks, that does seem like a legitimate problem, I agree. Rather than implementing this custom unmarshaling though I think we could make use of `mapstructure` and `WeaklyTypedInput` (https://pkg.go.dev/github.com/go-viper/mapstructure/v2#example-Decode-WeaklyTypedInput) which should massively reduce the amount of custom code we need to maintain here.

Thank you for the suggestion! I completely agree.
I'll go ahead and adapt my implementation to use this method and update the code accordingly.

File: tools.go [hunk 1]
```diff
@@ -70,6 +72,57 @@ func MustTool[T any, R any](
...
+	if err != nil {
+		return fmt.Errorf("failed to create decoder: %w", err)
+	}
+
+	if err := msDecoder.Decode(raw); err != nil {
+		return fmt.Errorf("failed to decode arguments: %w", err)
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (sd2k) [reply tools.go:158]: Bleh, this is true, and means `mapstructure` might not work 😞 sorry about that! I asked Claude for some ideas and it suggested this approach:

> Suggested approach: JSON byte preprocessing. Instead of decoding into a map and re-encoding, we can work at the json.RawMessage level: parse the JSON into map[string]json.RawMessage, strip the surrounding quotes from string values where the target struct expects an integer ("42" → 42), then let standard json.Unmarshal handle everything — type validation, overflow detection, custom unmarshalers, and all. This gets us
>   down to ~50 lines with no new dependencies:
> 
>   ```go
>   func unmarshalWithIntConversion(data []byte, target interface{}) error {
>         targetType := reflect.TypeOf(target)
>         if targetType.Kind() != reflect.Ptr || targetType.Elem().Kind() != reflect.Struct {
>                 return json.Unmarshal(data, target)
>         }
> 
>         intFields := collectIntFieldNames(targetType.Elem())
>         if len(intFields) == 0 {
>                 return json.Unmarshal(data, target)
>         }
> 
>         var raw map[string]json.RawMessage
>         if err := json.Unmarshal(data, &raw); err != nil {
>                 return err
>         }
> 
>         changed := false
>         for fieldName := range intFields {
>                 rawVal, ok := raw[fieldName]
>                 if !ok || len(rawVal) < 2 {
>                         continue
>                 }
>                 if rawVal[0] == '"' && rawVal[len(rawVal)-1] == '"' {
>                         raw[fieldName] = rawVal[1 : len(rawVal)-1]
>                         changed = true
>                 }
>         }
> 
>         if !changed {
>                 return json.Unmarshal(data, target)
>         }
> 
>         fixed, err := json.Marshal(raw)
>         if err != nil {
>                 return err
>         }
>         return json.Unmarshal(fixed, target)
>   }
>   ```
> 
>   Where collectIntFieldNames walks the struct type (including embedded structs) to build a set of JSON field names that map to integer types. The fast path (no int fields, or no strings needing conversion) falls through to a single json.Unmarshal with zero overhead.
> 
>   Could you update the PR with this approach? Happy to help if you have questions.
>

File: tools.go [hunk 2]
```diff
@@ -70,6 +70,173 @@ func MustTool[T any, R any](
...
+
+	// Reject JSON keywords that would be valid JSON but aren't numeric
+	str := string(data)
+	if str == "null" || str == "true" || str == "false" {
+		return false
+	}
+
+	// Check if it looks like a number: optional minus, followed by digits
+	i := 0
+	if data[i] == '-' || data[i] == '+' {
```
[the remark this answers is not in the document]
Comment (ioanarm) [reply tools.go]: I think this is valid?

File: go.mod [hunk 3]
```diff
@@ -80,7 +80,7 @@ require (
 	github.com/go-openapi/swag/typeutils v0.25.4 // indirect
 	github.com/go-openapi/swag/yamlutils v0.25.4 // indirect
 	github.com/go-openapi/validate v0.25.1 // indirect
-	github.com/go-viper/mapstructure/v2 v2.4.0 // indirect
+	github.com/go-viper/mapstructure/v2 v2.5.0 // indirect
```
Comment (ioanarm) [inline go.mod]: This is probably unnecessary now. it's not a problem but it's a bit of noise for this PR. Can you revert?

**Comment by ioanarm** [comment]
Hello @leozeli! When you have some time can you please resolve the conflict so we can have a final review on the PR? 
Thanks!

**Comment by leozeli** [comment]
> Hello @leozeli! When you have some time can you please resolve the conflict so we can have a final review on the PR?
> Thanks!

Will do! I'll fix the conflicts and update the PR shortly. Thanks!

File: tools_string_to_int_test.go
Comment (sd2k) [inline tools_string_to_int_test.go:1]: I suspect these tests could be a lot more succinct using table tests.

**Comment by sd2k** [comment]
@leozeli thanks for the fixes! I think our CLA Assistant has decided to start working again now, could you sign the CLA and I can get this in?

**Comment by leozeli** [comment]
> @leozeli thanks for the fixes! I think our CLA Assistant has decided to start working again now, could you sign the CLA and I can get this in?

Signed! Thanks for reviewing and getting this in.
